src/EditorFeatures/TestUtilities/CompleteStatement/AbstractCompleteStatementTests.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis.Editor.UnitTests.Utilities;
+using Microsoft.CodeAnalysis.Editor.UnitTests.Workspaces;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Roslyn.Test.Utilities;
+using Xunit;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.UnitTests.CompleteStatement
+{
+    [UseExportProvider]
+    public abstract class AbstractCompleteStatementTests
+    {
+        internal static char semicolon = ';';
+
+        internal abstract VSCommanding.ICommandHandler GetCommandHandler(TestWorkspace workspace);
+
+        protected abstract TestWorkspace CreateTestWorkspace(string code);
+
+        protected void VerifyTypingSemicolon(string initialMarkup, string expectedMarkup, string newLine = "\r\n")
```

Comment:
Could you desribe what the test harness is doing in broad terms?

Reply:
This method and the method above for "no special behavior" are now documented.